cloud/replication_test.cc

```diff
@@ -0,0 +1,601 @@
+
+  // Returns the number of log records applied
+  size_t catchUpFollower();
+
+  WriteOptions wo() const {
+    WriteOptions w;
+    w.disableWAL = true;
+    return w;
+  }
+
+  void createColumnFamily(std::string name);
+  void deleteColumnFamily(std::string name);
+
+ private:
+  std::string test_dir_;
+  FollowerEnv follower_env_;
+
+  port::Mutex log_records_mutex_;
+  std::vector<ReplicationLogRecord> log_records_;
+  int followerSequence_{0};
+
+  std::unique_ptr<DB> leader_db_;
+  ColumnFamilyMap leader_cfs_;
+  std::unique_ptr<DB> follower_db_;
+  ColumnFamilyMap follower_cfs_;
+};
+
+Options ReplicationTest::leaderOptions() const {
+  Options options;
+  options.create_if_missing = true;
+  options.atomic_flush = true;
+  options.avoid_flush_during_shutdown = true;
+  options.write_buffer_size = 100 << 10;
+  options.target_file_size_base = 500 << 10;
+  options.max_background_jobs = 4;
+  options.max_open_files = 500;
+  options.max_bytes_for_level_base = 1 << 20;
+  return options;
+}
+
+DB* ReplicationTest::openLeader() {
+  bool firstOpen = log_records_.empty();
+  auto dbname = test_dir_ + "/leader";
+
+  std::vector<std::string> cf_names;
+  auto s = DB::ListColumnFamilies(Options(), dbname, &cf_names);
+  assert(firstOpen == !s.ok());
+  if (!s.ok()) {
+    cf_names.push_back(kDefaultColumnFamilyName);
+  }
+
+  auto options = leaderOptions();
+  auto listener =
+      std::make_shared<Listener>(&log_records_mutex_, &log_records_);
+  options.replication_log_listener = listener;
+
+  listener->setState(firstOpen ? Listener::TAILING : Listener::OPEN);
+
+  std::vector<ColumnFamilyDescriptor> column_families;
+  for (auto& name : cf_names) {
+    column_families.emplace_back(name, options);
+  }
+
+  std::vector<ColumnFamilyHandle*> handles;
+  DB* db;
+  // open DB
+  s = DB::Open(options, dbname, column_families, &handles, &db);
+  assert(s.ok());
+  leader_db_.reset(db);
+  // Follower will sometimes need to access deleted files
+  db->DisableFileDeletions();
+
+  for (auto& h : handles) {
+    bool inserted = leader_cfs_.try_emplace(h->GetName(), h).second;
+    assert(inserted);
+  }
+
+  if (!firstOpen) {
+    MutexLock lock(&log_records_mutex_);
+    listener->setState(Listener::RECOVERY);
+    // recover leader
+    DB::ApplyReplicationLogRecordInfo info;
+    auto leaderSeq = getPersistedSequence(db) + 1;
+    for (; leaderSeq < (int)log_records_.size(); ++leaderSeq) {
+      s = db->ApplyReplicationLogRecord(log_records_[leaderSeq], &info);
+      assert(s.ok());
+    }
+    listener->setState(Listener::TAILING);
+  }
+
+  return db;
+}
+
+DB* ReplicationTest::openFollower() {
+  auto dbname = test_dir_ + "/follower";
+
+  std::vector<std::string> cf_names;
+  auto s = DB::ListColumnFamilies(Options(), dbname, &cf_names);
+  if (!s.ok()) {
+    cf_names.push_back(kDefaultColumnFamilyName);
+  }
+
+  auto options = leaderOptions();
+  options.env = &follower_env_;
+  options.disable_auto_compactions = true;
```

Comment:
We want the leader to execute compactions and propagate compaction results through the replication log into the MANIFEST. If the follower executed the compaction and wrote the changes to MANIFEST it would diverge from leader's MANIFEST.

We also have to disable follower's flush for the same reason.